CIP4/CIP4.md

```diff
@@ -0,0 +1,132 @@
+
+#### Letter part
+
+For letters, we render the bytes in hex, but replace the alphanumeric used in hex with this letter-only alphabet:
+
+`A B C D E J H K L N O P S T X Z`
+
+This alphabet satisfies the following requirements:
+
+1) Has exactly 16 letters (one-to-one mapping with 2 bytes in HEX)
+1) No contained characters that look too much like each other
+1) Minimizes occurrences of undesirable words in [this list](https://www.noswearing.com/fourletterwords.php).
+
+#### Number part
+
+The last two bytes are compressed to a 4-digit number. For this we will simply take the last 4 digits of the 16-bit integer number constructed from 2 bytes as `((A << 8) + B) % 10000` (zero-padded).
+
+This above produces 10000 unique values across all possible values of A and B and giving maximum of 7 potential collisions per value and 6.5 average collisions per value, which is the minimum, given the fact that we reduce maximum potential number 65025 to 4 digits.
+**Note**: resulting number is also zero-padded to 4 digits.
+
+### ImagePart rationale
+
+For the image, we take the result of `hash1` and use it as the seed for the [blockies](https://github.com/ethereum/blockies) library.
+
+This library in particular has the following benefits:
+
+- Has been audited
+- Used by other blockchains which and therefore has common libraries for implementation
+
+**Note**: This library internally re-hashes its input to a 128-bit entropy string
+
+### Hash algorithms used in Byron + ITN
+
+For `hash1`, we use `blake2b512`. [Blake2b](https://tools.ietf.org/html/rfc7693) is a standardized hash function that is used in Cardano for other purposes like key derivations. Reusing blake2b means one less dependency. We use `512` bytes of output to try and future-proof this algorithm (better to spread the entropy across more bits than needed than end up not capturing the full entropy in the future).
+
+For `hash_to_32` we use CRC32. We hash a second time for the following:
+
+1) The *TextPart* is constructed from 4 bytes (2 for letters, 2 for numbers) and so we need to project the result of `hash1` down to 4 bytes.
+2) We don't want to simply take the last 4 bytes of `hash1` because that would reveal part of the input used to generate the *ImagePart*. Although strictly speaking this should not be of a concern (since the result of `hash1` doesn't reveal any information about the original key), we take this as a precaution.
+3) `CRC32` is used in the Byron implementation of Cardano as a checksum for addresses, meaning no additional dependency has to be added.
+
+Although there is no specification for CRC32 and many variations exist, in Cardano we use the CRC-32-IEEE variation. You can find a C implementation [here](https://github.com/cardano-foundation/ledger-app-cardano/blob/3f784d23c1b87df73cda552ef01428d3e2733237/src/crc32.c#L6)
+
+### Hash algorithms used in Shelley mainnet
```

Comment:
This part is mostly the part up for discussion. I'm open to any change for Shelley (since all users will get a new checksum anyway), or we can decide to keep it exactly the same if people want that